apps/saleor-app-checkout/backend/payments/providers/adyen/utils.ts

```diff
@@ -44,12 +55,209 @@ export const getLineItems = (lines: OrderFragment["lines"]): Types.checkout.Line
 export const createEventUniqueKey = (event: TransactionEventFragment) =>
   [event.name, event.reference].join();
 
-export const getAmountAfterRefund = (money: MoneyFragment, refundAmount: number) => {
-  const chargedAmount = money.amount - refundAmount;
+export const encodeBasicAuth = (username: string, password: string) => {
+  return Buffer.from(username + ":" + password, "ascii").toString("base64");
+};
+
+export const verifyBasicAuth = (username: string, password: string, token: string) => {
+  const credentials = encodeBasicAuth(username, password);
+  return token === `Basic ${credentials}`;
+};
+
+export const getNotificationStatus = (
+  notification: Types.notification.NotificationRequestItem
+): TransactionStatus => {
+  const { eventCode, success } = notification;
+  if (
+    failedEvents.has(eventCode) ||
+    success === Types.notification.NotificationRequestItem.SuccessEnum.False
+  ) {
+    return "FAILURE";
+  }
+
+  return "SUCCESS";
+};
+
+export type TransactionAmounts = Pick<
+  TransactionUpdateInput,
+  "amountRefunded" | "amountAuthorized" | "amountCharged" | "amountVoided"
+>;
+
+const nonNegative = (amount: number) => (amount < 0 ? 0 : amount);
+
+const getCurrencyFromTransaction = (transaction: TransactionFragment | null) => {
+  if (!transaction) {
+    return undefined;
+  }
+
+  const currencies = [
+    transaction?.refundedAmount?.currency,
+    transaction?.voidedAmount?.currency,
+    transaction?.chargedAmount?.currency,
+    transaction?.authorizedAmount?.currency,
+  ];
+
+  // find first existing currency
+  return currencies.find((item) => typeof item === "string");
+};
+
+export const getTransactionAmountFromAdyen = (
```

Comment:
Looks like this function could be split into a few smaller ones. Right now it's hard for me to follow what's happening.

Reply:
I've tried to refactor this a big - especially the conditions to validate the notification and currency, but I don't really have a better idea how to improve the big chunk of `switch` statement. I guess we could refactor each condition into a separate function, but there wouldn't be any overlap in functionality (maybe with `CancelOrRefund` case) - each case needs to return different data.